### PR TITLE
[jit] Use a points-to graph for alias analysis

### PR DESCRIPTION
**Pull request by suo**
This PR changes the way we store aliasing information from a "set" approach to a "points-to" analysis. Set-based approaches lose information in ways that make it difficult to do "live" updates to the alias DB as one as mutating the graph. 

The tradeoff is that simple queries get more expensive, since they require traversing the points-to graph to answer most questions. In practice, this is unlikely to be that costly since we don't have massive aliasing chains, but we could create an approximation/caching layer if this becomes a problem.

My rough plan is:
1. This PR, switching to a points-to graph
2. Make it "live": analyzing a node should record all the edges the node added, so that we can rollback when the node is destroyed. 
3. Reduce wildcard scope: we can make the wildcard a special vertex that points to anything that we're not "sure" about; namely, things that have been put inside lists, or graph inputs.

### CODE REVIEW

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 1]
```diff
@@ -21,74 +23,329 @@ bool shouldAnnotate(const Value* v) {
...
+        [&](const Element* el) { return el->value == b; },
+        BfsDirection::POINTS_TO,
+        /*shortCircuit=*/true);
+  }
+
+  // Make `v` point at `to`.
+  //
+  // Returns true if this actually changed the AliasTracker state (i.e. if
+  // we've never heard about this relationship before)
+  bool makeAlias(const Value* v, const Value* to) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/alias_analysis.cpp]: The name of this function makes it sound symmetric, but its implementation in wildcard handling is not symmetric.

**Review by zdevito — APPROVED** [review]
I didn't see anything obviously wrong. I think this will look cleaner after wildcard handling is no longer special.

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 2]
```diff
@@ -21,74 +23,329 @@ bool shouldAnnotate(const Value* v) {
...
+    template <typename Fn>
+    bool bfsBothDirections(Fn fn, bool shortCircuit = false) const {
+      return this->bfs(fn, BfsDirection::POINTS_TO, shortCircuit) ||
+          this->bfs(fn, BfsDirection::POINTED_FROM, shortCircuit);
+    }
+  };
+
+  // Structure that owns all the element pointers. It's a map of
+  //  raw pointer -> unique_ptr to facilitate easy queries
+  std::unordered_map<Element*, std::unique_ptr<Element>> elements_;
```
[the remark this answers is not in the document]
Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: In the future when we want to delete an element we'll need some way of finding it by raw ptr.

File: torch/csrc/jit/passes/alias_analysis.cpp (hunk 1 above)
Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: I'll change it makePointerTo or something like that

**Pull request closed without merging**